src/building/bootstrapping.md

```diff
@@ -264,6 +264,55 @@ is built by the _beta_ compiler, but using the _master_ version of libstd!
 The only time `rustc` uses `cfg(bootstrap)` is when it adds internal lints
 that use diagnostic items. This happens very rarely.
 
+### What is a 'sysroot'?
+
+When you build a project with cargo, the build artifacts for dependendencies
+are normally stored in `target/debug/deps`. This only contains dependencies cargo
+knows about; in particular, it doesn't have the standard library. Where do
+`std` or `proc_macro` come from? It comes from the **sysroot**, a directory
+where the compiler loads build artifacts at runtime. The sysroot doesn't just
+store the standard library, though - it includes anything that needs to be
+loaded at runtime. That includes (but is not limited to):
+
+- `libstd`/`libtest`/`libproc_macro`
+- The compiler crates themselves, when using `rustc_private`. In-tree these
+  are always present; out of tree, you need to install `rustc-dev` with rustup.
+- `libLLVM.so`, the shared object file for the LLVM project. In-tree this is
+  either built from source or downloaded from CI; out-of-tree, you need to
+  install `llvm-tools-preview` with rustup.
+
+All the artifacts listed so far are *compiler* runtime dependencies. You can
+see them with `rustc --print sysroot`:
+
+```
+$ ls $(rustc --print sysroot)/lib
+libLLVM-11-rust-1.50.0-nightly.so    libtest-57470d2aa8f7aa83.so
+librustc_driver-4f0cc9f50e53f0ba.so  libtracing_attributes-e4be92c35ab2a33b.so
+librustc_macros-5f0ec4a119c6ac86.so  rustlib
```

Comment:
It's perhaps worth noting here that the lib directory is not part of search paths for linkers etc when building Rust code (ever) -- rustlib/$target/lib is, though. That is, the lib directory's contents can be considered an implementation detail of current Rust, but aren't really guaranteed to exist at all. For example, libtracing_attributes here is never going to be used, we could stop shipping it and no one would notice. The only reason it exists in Rust installs is filename-based filtering when copying artifacts.

Also, on an unrelated note - this list looks incomplete. It appears to be missing libstd? Not that it really matters and I might not even provide these lists...

```
libchalk_derive-0685d79833dc9b2b.so  librustc_driver-4f0cc9f50e53f0ba.so  libstd-25c6acf8063a3802.so   libtracing_attributes-e4be92c35ab2a33b.so
libLLVM-11-rust-1.50.0-nightly.so    librustc_macros-5f0ec4a119c6ac86.so  libtest-57470d2aa8f7aa83.so  rustlib
```

Reply:
> That is, the lib directory's contents can be considered an implementation detail of current Rust, but aren't really guaranteed to exist at all.

Isn't that true of both /lib and /lib/rustlib? I didn't think we guaranteed any part of the sysroot to be stable, unless you mean `rustlib` is de facto stable but `lib` isn't even that.

> It appears to be missing libstd? 

Good catch, thanks, I was missing a line.

Reply:
We could add stable crates to lib safely, i.e., without the rustc_private adding -Z flag, but not to rustlib is basically my point.

Reply:
> We could add stable crates to lib safely, i.e., without the rustc_private adding -Z flag, but not to rustlib is basically my point.

I don't understand what this means, sorry. What does `rustc_private` have to do with `-Z` flags? What does stability have to do with sysroots?

Reply:
rustc_private is enforced on all rustc dependencies via a -Z flag, force-unstable-if-unmarked=rustc_private IIRC. This was (more) necessary at some point in the past when the rustc deps where shipped to all Rust users; these days given you need to add rustc-dev it is somewhat less needed.

Crate artifacts placed in rustlib/$target/lib are essentially stabilized (not location but existence) unless they're feature gated carefully (e.g., via the force-unstable-if-unmarked flag). This is not true of the lib directory, where we could place any artifact we wanted but it would be essentially unreachable to Rust code.

Reply:
I'm still slightly confused - don't you need `#![feature(rustc_private)]` to load things from `rustlib/$target/lib`? Does `force-unstable-if-unmarked` automatically add that feature gate to all the rustc crates when they're compiled with x.py?

>  This is not true of the lib directory, where we could place any artifact we wanted but it would be essentially unreachable to Rust code.

That bit makes sense to me, since `lib/` isn't in the search path for the linker.

Reply:
No, you don't need that feature gate to load a crate from sysroot (otherwise std would be feature gated!); you need the feature attribute if you're loading a feature gated crate. All crates compiled during rustc build are marked unstable unless they explicitly stabilize themselves via the -Z flag.

Reply:
I updated this to mention `-Z force-unstable-if-unmarked`. Thanks for explaining it!